File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    string buttons[10] = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    vector<string> letterCombinations(string digits)
    {
        vector<string> result;
        dfs(0, "", digits, result);
        
        return result;
    }
    
    void dfs(int start, string s, string digits, vector<string>& result)
    {
        if(start == digits.length() && s.length() == digits.length() && s.length() != 0)
        {
            result.push_back(s);
            return;
        }
        
        for(int i = start; i < digits.length(); i++)
        {
            int j = stoi(digits.substr(i, 1));
            string button = buttons[j];
            
            for(auto c : button)
            {
                dfs(i + 1, s + c, digits, result);
            }
        }
    }
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    hashmap = {"2":"abc", "3":"def", "4":"ghi", "5":"jkl","6":"mno", "7":"pqrs","8":"tuv","9":"wxyz"}
    
    
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []
        
        if len(digits) == 1:
            return [x for x in self.hashmap[digits[0]]]

        mid = len(digits) // 2
        
        left = self.letterCombinations(digits[:mid])
        right = self.letterCombinations(digits[mid:])
        
        ans = []
        for l in left:
            for r in right:
                ans.append(l + r)
        
        return ans